### PR TITLE
Allow the specification of the self keyword in the target of an sesea…

### PR DESCRIPTION
**Pull request by dburgener**
…rch query

This will return rules where the source type and target type are the same

### CODE REVIEW

File: sesearch [hunk 1]
```diff
@@ -144,6 +144,9 @@ try:
     p = setools.SELinuxPolicy(args.policy)
 
     if args.tertypes:
+        self_rule = args.target == "self"
+        if self_rule:
+            args.target = None
```
Comment (pebenito) [inline sesearch:149]: I'd rather have an explicit `--self` option that conflicts with the `-t` option (see the above checks that call parser.error() for example). It could probably conflict with the regex and indirect options for the target too, though I'm not sure that needs to be checked.

File: setools/terulequery.py [hunk 2]
```diff
@@ -162,13 +165,21 @@ def results(self):
             #
             # Matching on target type
             #
-            if self.target and not match_indirect_regex(
+            if self.target and not self.target_self and not \
```
Comment (pebenito) [inline setools/terulequery.py:168]: There should also be a check after the debug messages above that raises and exception if `self.target` and `self.target_self` are set, as they conflict. That will simplify logic down here.

There should be a warning level log message for using target_regex or target_indirect with target_self since the former are ignored with self.

File: setools/terulequery.py [hunk 3]
```diff
@@ -162,13 +165,21 @@ def results(self):
...
                     self.target_indirect,
                     self.target_regex):
                 continue
 
+            if self.target_self and not match_indirect_regex(
+                    rule.target,
+                    rule.source,
+                    False,
+                    False):
+                continue
```
Comment (pebenito) [inline setools/terulequery.py:181]: Please use explicit `rule.target == rule.source` check here so it is crystal clear.

**Review by pebenito — CHANGES_REQUESTED** [review]
I'd also like to see at least 2 unit tests, minimum test is that it works right when the source is a type and when it is an attribute. Extra credit if you check the exception path for setting target and target_self on the query class.

File: setools/terulequery.py [hunk 4]
```diff
@@ -162,13 +172,17 @@ def results(self):
             #
             # Matching on target type
             #
-            if self.target and not match_indirect_regex(
+            if self.target and not \
+                    match_indirect_regex(
```
Comment (pebenito) [inline setools/terulequery.py:176]: Please drop this change.

File: setools/terulequery.py [hunk 5]
```diff
@@ -162,13 +172,17 @@ def results(self):
...
                     rule.target,
                     self.target,
                     self.target_indirect,
                     self.target_regex):
                 continue
 
+            if self.target_self and not rule.target == rule.source:
+                continue
```
Comment (pebenito) [inline setools/terulequery.py:184]: On further thought, I think this is too simplistic in cases of attributes, particularly for `allow attr attr:...`.

If you do `allow attr self:...` that expands the rule into `allow type1 type1:... allow type2 type2:...`. `allow attr attr` does not correspond to a source policy self rule. I'm inclined to say we don't want `allow attr attr` to match since that doesn't match a self rule as written in source, but I am open to arguments.

@stephensmalley @jwcart2 @jbrindle @doverride @cgzones @bauen1 any opinions?

**Review by pebenito — CHANGES_REQUESTED** [review]

File: setools/terulequery.py (hunk 5 above)
Comment (dburgener) [reply setools/terulequery.py:184]: In my opinion, I think of sesearch as being more fundamentally about permission search than rule search.

If I run sesearch for "allow type1 self", and "allow attr attr" is in the policy, then the permission I searched for is present, even though the specific rule is not.  That's information I would like to get in that query.

Comment (bauen1) [reply setools/terulequery.py:184]: > If you do allow attr self:... that expands the rule into allow type1 type1:... allow type2 type2:.... allow attr attr does not correspond to a source policy self rule. I'm inclined to say we don't want allow attr attr to match since that doesn't match a self rule as written in source, but I am open to arguments.

I would argue that the `self` keyword is syntactic sugar to indicate the same target type as the source type.
`(allow a a (.))` and `(allow a self (.))` are the same, there for a search for allow rules with the source type `a` with either the same target type `a` or the syntactic sugar `self` should return both rules (maybe formatted accordingly to the query).

Comment (pebenito) [reply setools/terulequery.py:184]: > `(allow a a (.))` and `(allow a self (.))` are the same

This is true where `a` is a type, but not if it is an attribute.

Comment (jbrindle) [reply setools/terulequery.py:184]: > I would argue that the self keyword is syntactic sugar to indicate the same target type as the source type.

self is not syntactic sugar, it is stored in the kernel policy and can express something not possible otherwise. Namely, when the source is an attribute, a self target means each type in the attribute as the source and itself as the target. For example, these are not equivalent:
allow domain domain:process fork;
allow domain self:process fork;

The first gives every domain process:fork for every domain, the latter gives every domain process:fork only on itself.

Comment (bauen1) [reply setools/terulequery.py:184]: Oh my brain just read what I wanted to read ...

> If you do allow attr self:... that expands the rule into allow type1 type1:... allow type2 type2:.... allow attr attr does not correspond to a source policy self rule. I'm inclined to say we don't want allow attr attr to match since that doesn't match a self rule as written in source, but I am open to arguments.

sesearch should output all rules that match `(allow a a (.))`, even in the form of typeattributes.
But (if possible, i.e. not if they have been expanded) the type attribute rules shouldn't use the `self` keyword when displaying the rule.

File: setools/terulequery.py (hunk 4 above)
Comment (dburgener) [reply setools/terulequery.py]: dropped

File: setools/terulequery.py (hunk 5 above)
Comment (bauen1) [reply setools/terulequery.py:183]: > self is not syntactic sugar
Is that true for both type attributes and normal types ?

File: setools/terulequery.py [hunk 6]
```diff
@@ -141,6 +144,13 @@ def results(self):
         self.log.debug("Boolean: {0.boolean!r}, eq: {0.boolean_equal}, "
                        "regex: {0.boolean_regex}".format(self))
 
+        if self.target and self.target_self:
+            raise ValueError("target and target_self both specified")
```
Comment (pebenito) [inline setools/terulequery.py:148]: I think there should be a new exception and it should probably subclass RuntimeError (and SEToolsException) specifically for signalling conflicting options.

File: setools/terulequery.py [hunk 7]
```diff
@@ -141,6 +144,13 @@ def results(self):
...
+
+        if self.target_self and (self.target_regex or (self.target_indirect is False)):
```
Comment (pebenito) [inline setools/terulequery.py:150]: Shouldn't this be for either target_regex or target_indirect are true, not false for the latter?

Comment (pebenito) [reply setools/terulequery.py:150]: Maybe it should just be for target_regex, since neither indirect setting makes sense for self.

Comment (dburgener) [reply setools/terulequery.py:150]: I believe that target_indirect defaults to True, so checking against (target_regex == True or target_indirect == False) would catch whether either of them are changed from their defaults.  Maybe I'm misunderstanding something about how the current code works?

Comment (pebenito) [reply setools/terulequery.py:150]: Your understanding is correct. Ok, lets keep the check but it should be more Pythonic:
`if self.target_self and (self.target_regex or not self.target_indirect):`

File: setools/terulequery.py (hunk 5 above)
Comment (pebenito) [reply setools/terulequery.py:183]: I have to stick with my original thought: `--self` not match in `allow attr attr:...` statements.

Comment (dburgener) [reply setools/terulequery.py:183]: I'm having a hard time seeing the value of this change with that.  Do you have any example of a motivating use case where someone might use this feature with the behavior of --self not matching "allow attr attr"?

I get where you're coming from in terms of matching the kernel rules, but my motivating example to want this is a situation where I want to know if a given domain has certain self access. For example if I want to know which domains allow anonymous pipes, I might run:

`sesearch -A -c fifo_file --self`

If --self doesn't match "allow attr attr", then a policy with something like "allow domain domain:fifo_file rw_fifo_file_perms;" would not match.  That means that I can't assume that the above query gives me all domains with fifo_file access.  It's true that it doesn't match the policy rule perfectly (and that "allow domain self" is very different access), but I'm no longer sure what scenarios I would use this query in if there are potentially permissions that are specified using attributes.

Comment (pebenito) [reply setools/terulequery.py:183]: I'd like to get this merged, so lets make `--self` sensitive to `self.target_indirect`. Then the user can get either behavior.
